tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs

```diff
@@ -281,6 +291,251 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
+			if (nsobjectConstructor is null && ctor is null)
+				return;
+
+			var createInstanceMethod = type.AddMethod ("_Xamarin_ConstructINativeObject", MethodAttributes.Public | MethodAttributes.Static | MethodAttributes.NewSlot | MethodAttributes.HideBySig, abr.ObjCRuntime_INativeObject);
+			var nativeHandleParameter = createInstanceMethod.AddParameter ("nativeHandle", abr.ObjCRuntime_NativeHandle);
+			var ownsParameter = createInstanceMethod.AddParameter ("owns", abr.System_Boolean);
+			abr.INativeObject__Xamarin_ConstructINativeObject.Resolve ().IsPublic = true;
+			createInstanceMethod.Overrides.Add (abr.INativeObject__Xamarin_ConstructINativeObject);
+			var body = createInstanceMethod.CreateBody (out var il);
+
+			if (nsobjectConstructor is not null) {
+				// var instance = new TypeA (nativeHandle);
+				// // alternatively with a cast: new TypeA ((IntPtr) nativeHandle);
+				// if (instance is not null && owns)
+				//     Runtime.TryReleaseINativeObject (instance);
+				// return instance;
+
+				if (type.HasGenericParameters) {
+					nsobjectConstructor = type.CreateMethodReferenceOnGenericType (nsobjectConstructor, type.GenericParameters.ToArray ());
+				}
+
+				il.Emit (OpCodes.Ldarg, nativeHandleParameter);
+				if (nsobjectConstructor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
+					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
+				il.Emit (OpCodes.Newobj, nsobjectConstructor);
+
+				var falseTarget = il.Create (OpCodes.Nop);
+				il.Emit (OpCodes.Dup);
+				il.Emit (OpCodes.Ldnull);
+				il.Emit (OpCodes.Cgt_Un);
+				il.Emit (OpCodes.Ldarg, ownsParameter);
+				il.Emit (OpCodes.And);
+				il.Emit (OpCodes.Brfalse_S, falseTarget);
+
+				il.Emit (OpCodes.Dup);
+				il.Emit (OpCodes.Call, abr.Runtime_TryReleaseINativeObject);
+
+				il.Append (falseTarget);
+
+				il.Emit (OpCodes.Ret);
+			} else if (ctor is not null) {
+				// return new TypeA (nativeHandle, owns); // for NativeHandle ctor
+				// return new TypeA ((IntPtr) nativeHandle, owns); // IntPtr ctor
+
+				if (type.HasGenericParameters) {
+					ctor = type.CreateMethodReferenceOnGenericType (ctor, type.GenericParameters.ToArray ());
+				}
+
+				il.Emit (OpCodes.Ldarg, nativeHandleParameter);
+				if (ctor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
+					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
+				il.Emit (OpCodes.Ldarg, ownsParameter);
+				il.Emit (OpCodes.Newobj, ctor);
+				il.Emit (OpCodes.Ret);
+			} else {
+				throw new UnreachableException ();
+			}
+
+			Annotations.Mark (createInstanceMethod);
+		}
+
+		static MethodReference? FindNSObjectConstructor (TypeDefinition type)
+		{
+			return FindConstructorWithOneParameter ("ObjCRuntime", "NativeHandle")
+				?? FindConstructorWithOneParameter ("System", "IntPtr");
+
+			MethodReference? FindConstructorWithOneParameter (string ns, string cls)
+				=> type.Methods.FirstOrDefault (method =>
```

Comment:
This seems safer, since in case the selection logic is wrong and matches more than one constructor:

```suggestion
				=> type.Methods.SingleOrDefault (method =>
```